### ISSUE TITLE
Support configuration of Solr API endpoint

### ISSUE THREAD

**Issue by nichtich** [pkiraly/qa-catalogue-web#109]
Solr instance is hard coded to run at localhost. This should be configurable.

Existing references to Solr API URL:

~~~php
$solrPath = $this->getIndexName();

$baseUrl = 'http://localhost:8983/solr/' . $solrPath; 
    $url = $baseUrl . '/select/?q=*:*&wt=csv&rows=0';

$url = 'http://localhost:8983/solr/' . $solrPath . '/select?' . join('&', $this->encodeSolrParams($params));

protected function isCoreAvailable($core) {
   $url = 'http://localhost:8983/solr/' . $core . '/admin/ping';

$url = 'http://localhost:8983/solr/' . $solrPath . '/select?' . $params;
$url = sprintf('http://localhost:8983/solr/%s/select?q=%s&fl=id&start=%d&rows=%d',
               $core, urlencode($query), $start, $rows)
~~~

I don't fully get the difference between "index name" and "core". There seem to be [three solr cores for each catalogue](https://github.com/pkiraly/qa-catalogue-web/blob/main/classes/Data.php#L430) (`$name`, `validation` and `${name}_validation`) plus the `${name}_dev`.

This is required by https://github.com/pkiraly/qa-catalogue/issues/289

**Comment by pkiraly** [comment]
I agree, it should not be hard coded. "index name" and "core" are the same thing.
The different between `$name` and `${name}_dev` is the same: indexing takes lots of time. It starts with clearing the index (removing all entries (in Solr they are called documents). At the end of index "commit" and "optimise" happens which finalizes the index. During the process it is not possible to search, or it could be done, but the process would take longer, and the index does not contains all bibliographic records. So the index happens in a `dev` version, so it does not work on the production index, but swap the two when it is ready. I think the same should happen with the other analyses, making the output all the time coherent.
`validation` is different, and needs some further tweaking. When working with K10plus, we needed to display the completeness and validation results by libraries. This requires some aggregation of results, which could be done with SQL `GROUP BY` commands, but it turned out that both sqlite3 and MySQL are slow. Solr's faceted search provides a solution, but it also turned out that updating an index is quite slow, and usually validation happens before Solr indexing, at a time when the Solr index is not available. So I created a new index to solve this problem. It contains record id, group ids (the ids of the libraries) and the ids of the validation errors. It is used by the validation sqlite phase, but if the data is already there and available, the Solr index copy this info as well. That would make the validation index out of use, since the main index also contains the same information. However when you want to rerun the validation, the index might be different (because you might change the code or the configuration of the validation), so the information stored in the main index is out of date.
Conclusion: either we make validation core temporary, but then we always run indexing after each validation, or we keep it semi-temporary: an index process always deletes the validation core, but the front end check if validation is available and use it when it is available because it is a sign that a new validation process happened after the last index, so that version might be better than the main index.

**Comment by nichtich** [comment]
As far as I understand `${name}_dev` is only used by the backend, while qa-catalogue-web does not even know that this additional index exists and it should run happily on a single index

> It is used by the validation sqlite phase, but if the data is already there and available, the Solr index copy this info as well. However when you want to rerun the validation, the index might be different (because you might change the code or the configuration of the validation), so the information stored in the main index is out of date.

So qa-catalogue-web [checks if the main index contains validation information](https://github.com/pkiraly/qa-catalogue-web/blob/main/classes/Data.php#L491-L495) and uses the main index `$name` only in this case. Otherwise it [checks whether index `validation` exists or if not `${name}_validation`](https://github.com/pkiraly/qa-catalogue-web/blob/main/classes/Data.php#L430), right?

My only concern is the generic index name `validation` seems problematic because multiple catalogues in the same Solr instance may clash. Can we just take `${name}_validation`?

> Conclusion: either we make validation core temporary, but then we always run indexing after each validation, or we keep it semi-temporary: an index process always deletes the validation core, but the front end check if validation is available and use it when it is available because it is a sign that a new validation process happened after the last index, so that version might be better than the main index.

As far as I understand the code, validation index is not used when available but when no error ids are in main index, so updating the validation only would not show ab in the web interface unless validation is removed from main index (?).

Seems like we have two modes of operation:

A) everything in one index
B) validation in a separate index

If validation index is needed anyway during analysis, why not always use mode B and never put validation into the main index?

**Comment by nichtich** [comment]
To get back to the original issue: Solr API can be configured with just one variable `SOLR`, having default value `http://localhost:8983/solr/`. The index name is from another configuration variable `indexName[<catalogue>]`.

**Comment by pkiraly** [comment]
> As far as I understand `${name}_dev` is only used by the backend
Yes, true, we do not need config for _dev

> My only concern is the generic index name validation seems problematic
I agree, this is just a leftover, I will delete it. The only really used one is `${name}_validation`

> If validation index is needed anyway during analysis, why not always use mode B and never put validation into the main index?
Because we can combine the validation result with other information in the search (e.g. "find all repeated subfields error in books published before 1820 in Latin" or "make a facet out of the validation errors"). We do not do that now, but on the long run it would be useful. In the long run I would like to put all the analysis result into the index.

**Comment by pkiraly** [comment]
> To get back to the original issue: Solr API can be configured with just one variable `SOLR`, having default value `http://localhost:8983/solr/`. The index name is from another configuration variable `indexName[<catalogue>]`.

I never used SolrCloud. If one config variable is OK for both indices I am fine with it.

**Comment by pkiraly** [comment]
Estimation: 3 hours

**Comment by pkiraly** [comment]
I have introduced two configuration properties:

- `mainSolrEndpoint`, `mainSolrEndpoint[<catalogue>]`: the URL of the main Solr endpoint. Default value: 
  `http://localhost:8983/solr/`. In multi-tenant mode you can specify it for a particular catalogue with
 `mainSolrEndpoint[<catalogue>]`.
- `solrEndpoint4ValidationResults`, `solrEndpoint4ValidationResults[<catalogue>]`: the URL of the Solr core that is 
  used for storing the results of validation. Usually this index is merged to the main core, so this property is needed
  on the special case when the validation is not merged. Default value: `http://localhost:8983/solr/`. 
  In multi-tenant mode you can specify it for a particular catalogue with `mainSolrEndpoint[<catalogue>]`.
